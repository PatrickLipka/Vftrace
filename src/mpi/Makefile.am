SUBDIRS = utils

if WITH_MPI
SUBDIRS +=vftr_mpi

SUBDIRS += c2vftr \
	   c-wrapper
if ENABLE_FORTRAN
SUBDIRS += f2vftr \
	   f-wrapper
if ENABLE_FORTRAN08
SUBDIRS += f082vftr \
	   f08-wrapper
endif
endif
endif
