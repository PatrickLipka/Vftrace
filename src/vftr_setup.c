/*
   This file is part of Vftrace.

   Vftrace is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   Vftrace is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#define _GNU_SOURCE

#include <fcntl.h>
#include <limits.h>
#include <string.h>
#include <sys/stat.h>
#include <assert.h>
#include <unistd.h>

#include "vftr_symbols.h"
#include "vftr_scenarios.h"
#include "vftr_hwcounters.h"
#include "vftr_version.h"
#include "vftr_environment.h"
#include "vftr_setup.h"
#include "vftr_dlopen.h"
#include "vftr_timer.h"
#include "vftr_fileutils.h"
#include "vftr_filewrite.h"
#include "vftr_browse.h"
#include "vftr_signals.h"
#include "vftr_stacks.h"
#include "vftr_hooks.h"
#include "vftr_timer.h"
#include "vftr_functions.h"
#include "vftr_allocate.h"

bool vftr_timer_end;

int vftr_mpirank;
int vftr_mpisize;
unsigned int vftr_function_samplecount;
unsigned int vftr_message_samplecount;

char *vftr_start_date;
char *vftr_end_date;

void vftr_print_disclaimer_full (FILE *fp) {
    fprintf (fp, 
        "\nThis program is free software; you can redistribute it and/or modify\n"
        "it under the terms of the GNU General Public License as published by\n"
        "the Free Software Foundation; either version 2 of the License , or\n"
        "(at your option) any later version.\n\n"
        "This program is distributed in the hope that it will be useful,\n"
        "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
        "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
        "GNU General Public License for more details.\n\n"
        "You should have received a copy of the GNU General Public License\n"
        "along with this program. If not, write to\n\n"
        "   The Free Software Foundation, Inc.\n"
        "   51 Franklin Street, Fifth Floor\n"
        "   Boston, MA 02110-1301  USA\n\n" );
}

/**********************************************************************/

void vftr_print_disclaimer (FILE *fp) {
    int v_major = MAJOR_VERSION;
    int v_minor = MINOR_VERSION;
    int rev = REVISION;
    fprintf (fp, "Vftrace version %d.%d.%d\n", v_major, v_minor, rev);
    fprintf (fp, "Runtime profile for application: %s\n", "");
    fprintf (fp, "Start Date: %s\n", vftr_start_date); 
    fprintf (fp, 
        "This is free software with ABSOLUTELY NO WARRANTY.\n"
        "For details: use vftrace with environment variable VFTR_LICENSE\n"
        "set to 1, or run \"vfview -w\", or consult the COPYRIGHT file.\n" );
}

/**********************************************************************/

void vftr_get_mpi_info (int *rank, int *size) {
#ifdef _MPI
// At this point, MPI_Init has not been called yet, so we cannot
// use MPI_Comm_size or MPI_Comm_rank. Instead, we have to rely
// on these environment variables set by various MPI implementations.
    if (getenv ("PMI_RANK")) {
        *rank = atoi (getenv("PMI_RANK"));
        *size = atoi (getenv("PMI_SIZE"));
    } else if (getenv ("OMPI_COMM_WORLD_RANK")) {
        *rank = atoi (getenv("OMPI_COMM_WORLD_RANK"));
	*size = atoi (getenv("OMPI_COMM_WORLD_SIZE"));
    } else if (getenv ("PMI_ID")) {
        *rank = atoi (getenv("PMI_ID"));
	*size = atoi (getenv("MPIRUN_NPROCS"));
    } else if (getenv ("MPIRUN_RANK")) {
        *rank = atoi (getenv ("MPIRUN_RANK"));
	*size = atoi (getenv ("MPIRUN_NPROCS"));
    } else if (getenv ("MPIRANK")) {
        *rank = atoi (getenv ("MPIRANK"));
        /* MPISIZE not set by MPI/SX, will be set after mpi_init call */
	*size = atoi (getenv ("MPISIZE"));
    } else if (getenv ("SLURM_PROCID")) {
	*rank = atoi (getenv ("SLURM_PROCID"));
	char *s;
	if (s = getenv ("SLURM_NPROCS")) *size = atoi (s);
    } else {
	// Cannot find out how many MPI ranks there are, assume only one.
	*rank = 0;
	*size = 1;
    }
#else
    // No MPI, only one rank exists.
    *rank = 0;
    *size = 1;
#endif
}
	
/**********************************************************************/

void vftr_set_start_date () {
   time_t t;
   time(&t);
   vftr_start_date = strdup(ctime (&t));
}

void vftr_set_end_date () {
   time_t t;
   time(&t);
   vftr_end_date = strdup(ctime (&t));
}

/**********************************************************************/

void vftr_initialize() {
    // set the timer reference point for this process
    vftr_set_local_ref_time();
    vftr_set_start_date();
    
    // measure overhead induced by vftrace
    long long overhead_time_start = vftr_get_runtime_usec();

    vftr_read_environment ();
    if (vftr_off()) {
	return;
    }
    vftr_assert_environment ();
	
    lib_opened = 0;

    vftr_timelimit = LONG_MAX;

    // No buffering for messages going directly to stdout
    setvbuf (stdout, NULL, _IOLBF, (size_t)0);

    sprintf (vftr_fileid, "VFTRACE %07d", 
	(100 * MAJOR_VERSION + MINOR_VERSION) * 100 + REVISION);

    vftr_overhead_usec = 0ll;
  
    vftr_prog_cycles = 0ll;

    vftr_program_path = vftr_get_program_path ();
    vftr_get_mpi_info (&vftr_mpirank, &vftr_mpisize);

    vftr_logfile_name = vftr_create_logfile_name (vftr_mpirank, vftr_mpisize, "log");

    vftr_log = fopen (vftr_logfile_name, "w+");
    assert (vftr_log);
    // Do not buffer when writing into the log file
    setvbuf (vftr_log, NULL, _IOLBF, (size_t)0);

    if (vftr_mpirank == 0) {
       if (vftr_environment.license_verbose->value) {
	  vftr_print_disclaimer_full (vftr_log);
       } else {
	  vftr_print_disclaimer (vftr_log);
       }
    }

    if (vftr_create_symbol_table (vftr_mpirank, NULL)) {
	// No symbol table has been found. Switch of symbol table.
	printf ("Vftrace could not find any parseable symbol tables associated with your executable\n");
	printf ("It will not be active for this application run\n");
	printf ("Consider recompiling Vftrace using -D__VMAP_OFFSET\n");
	vftr_switch_off();	
    }

    if (vftr_off()) {
        vftr_events_enabled = false;
    }

    memset (&vftr_prof_data, 0, sizeof(profdata_t));

    // initialize the stack variables and tables
    vftr_initialize_stacks();

    vftr_nextsampletime = 0ll;
    vftr_prevsampletime = 0;

    /* Init event counters */
    vftr_n_hw_obs = 0;
    if (vftr_environment.scenario_file->set) {
	vftr_events_enabled = !vftr_init_hwc (vftr_environment.scenario_file->value);
    } else {
	vftr_events_enabled = false;
    }
    if (vftr_n_hw_obs < 0) {
        fprintf(vftr_log, "error initializing H/W counters\n");
        vftr_events_enabled = false;
    }

    if (vftr_n_hw_obs  > 0) {
       vftr_prof_data.events[0] = (long long *) malloc (vftr_n_hw_obs * sizeof(long long));
       vftr_prof_data.events[1] = (long long *) malloc (vftr_n_hw_obs * sizeof(long long));
       memset (vftr_prof_data.events[0], 0, vftr_n_hw_obs * sizeof(long long));
       memset (vftr_prof_data.events[1], 0, vftr_n_hw_obs * sizeof(long long));
    }

    vftr_initcycles = vftr_get_cycles();
    // convert the sampletime and timelimit to microseconds
    vftr_interval  = (long long) (vftr_environment.sampletime->value * 1.0e6);
    assert (vftr_interval > 0ll);
    vftr_timelimit = vftr_environment.stoptime->value * 1000000ll;

    vftr_sorttime = 15.;
    vftr_sorttime_growth = 2.;

    /* Create VFD files */
    if (vftr_env_do_sampling ()) {
	vftr_init_vfd_file ();
    }
    
    vftr_profile_wanted = (vftr_environment.logfile_all_ranks->value) || (vftr_mpirank == 0);

    if (vftr_environment.print_stacks_for->set) {
        char *vftr_print_groups = vftr_environment.print_stacks_for->value;
        while (*vftr_print_groups && !vftr_profile_wanted) {
            int group_base, group_size;
            char *p;
	    // Loop to the end of the group, indicated by ","
            for (p = vftr_print_groups; *p && *p != ','; p++);
	    // sscanf returns the number of variables which have been filled
            switch (sscanf (vftr_print_groups, "%d:%d", &group_base, &group_size))  {
            case 1:
	      // Only one rank is to be printed. Check if it is the own rank.
              vftr_profile_wanted = vftr_mpirank == group_base;
	      break;
            case 2:
	      // Check if the own rank is in the interval given by [group_base, group_base + group_size).
              vftr_profile_wanted = vftr_mpirank >= group_base &&
                                    vftr_mpirank <  group_base + group_size;
            }
	    // Check if there is anything behind the comma of the environment variable. If so, proceed.
            vftr_print_groups = *p ? p + 1 : p;
        }
    }

    /* Define signal handlers */
    if (!vftr_environment.signals_off->value) {
	vftr_define_signal_handlers ();
    }

    fflush (stdout);
    vftr_initcycles = vftr_get_cycles();
    
    // get the time to estimate vftrace overhead
    long long overhead_time_end = vftr_get_runtime_usec();
    vftr_overhead_usec += overhead_time_end - overhead_time_start;
}

/**********************************************************************/

void vftr_calc_tree_format (function_t *func) {
    long long   fcalls, ftime;

    if (func == NULL) return;

    fcalls = func->prof_current.calls;
    ftime  = func->prof_current.cycles;
    if (vftr_maxtime < ftime) vftr_maxtime = ftime;

    int n = func->levels;

    /* Recursive search of callees */
    int i;
    function_t *f;
    for (i = 0, f = func->first_in_level; i < n; i++,f = f->next_in_level) {
        vftr_calc_tree_format (f);
    }
}

/**********************************************************************/

void vftr_finalize() {
    int ntop = 0;
    function_t **funcTable;

    if (vftr_off())  return;
    vftr_set_end_date();

    // get the total runtime
    long long finalize_time = vftr_get_runtime_usec();

    vftr_timer_end = true;

    // Mark end of non-parallel interval
    if (vftr_env_do_sampling()) {
        vftr_write_to_vfd (finalize_time, NULL, NULL, 0, SID_EXIT);
    }

    if (vftr_environment.strip_module_names->value) {
	vftr_strip_all_module_names ();
    }
    
    vftr_normalize_stacks();
    vftr_calc_tree_format (vftr_froots);

    display_function_t **display_functions;
    int n_display_functions;
    //if (vftr_environment.print_stack_profile->value || vftr_environment.create_html->value) {
       display_functions = vftr_create_display_functions (vftr_environment.mpi_show_sync_time->value, &n_display_functions, true); 
    //}

    FILE *f_html = NULL;
    if (vftr_environment.create_html->value) {
       vftr_browse_create_directory ();
       f_html = vftr_browse_init_profile_table (display_functions, n_display_functions);
    }

    //vftr_print_profile (vftr_log, display_functions, n_display_functions, &ntop, vftr_get_runtime_usec());
    vftr_print_profile (vftr_log, f_html, &ntop, vftr_get_runtime_usec());
#ifdef _MPI
    ////if (vftr_environment.print_stack_profile->value) {
       // Inside of vftr_print_function_statistics, we use an MPI_Allgather to compute MPI imbalances. Therefore,
       // we need to call this function for every rank, but give it the information of vftr_profile_wanted
       // to avoid unrequired output.
       vftr_print_function_statistics (vftr_log, display_functions, n_display_functions, vftr_profile_wanted);
    ////}
#endif
 
    funcTable = vftr_func_table;

    if (vftr_profile_wanted) {
        vftr_print_global_stacklist(vftr_log);
    }

    vftr_finalize_vfd_file (finalize_time, vftr_signal_number);
    if (vftr_events_enabled && vftr_stop_hwc() < 0) {
	fprintf(vftr_log, "error stopping H/W counters, ignored\n");
    }

    if (vftr_max_allocated_fields > 0) vftr_allocate_finalize(vftr_log);

    if (vftr_environment.print_env->value) vftr_print_environment(vftr_log);
    
    if (vftr_log) {
    	bool is_empty = (ftello (vftr_log) == (off_t)0);
    	if (is_empty) unlink (vftr_logfile_name);

    	fclose (vftr_log);
    }
    vftr_switch_off();
}

/**********************************************************************/

// vftr_finalize has to be called in the wrapper of MPI_Finalize, both for C and Fortran.
// This is the corresponding symbol for Fortran, with an added "_".
void vftr_finalize_() {
	vftr_finalize();
#ifdef _MPI
	PMPI_Barrier (MPI_COMM_WORLD);
#endif
}

/**********************************************************************/

int vftr_setup_test_1 (FILE *fp) {
	fprintf (fp, "Check MPI rank and size received from environment variables\n");		
	int mpi_rank, mpi_size;
	vftr_get_mpi_info (&mpi_rank, &mpi_size);
	fprintf (fp, "Rank: %d\n", mpi_rank);
	fprintf (fp, "Size: %d\n", mpi_size);
	return 0;
}

/**********************************************************************/

int vftr_setup_test_2 (FILE *fp) {
	fprintf (fp, "Check disclaimers\n");
	vftr_print_disclaimer_full (fp);
	fprintf (fp, "****************************************\n");
	vftr_print_disclaimer (fp);
	return 0;
}

