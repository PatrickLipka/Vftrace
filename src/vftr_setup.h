/*
   This file is part of Vftrace.

   Vftrace is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   Vftrace is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
#ifndef VFTR_SETUP_H
#define VFTR_SETUP_H

#include <stdio.h>
#include <stdbool.h>

extern bool vftr_timer_end;

extern int vftr_mpirank;
extern int vftr_mpisize;
extern unsigned int vftr_function_samplecount;
extern unsigned int vftr_message_samplecount;

extern bool vftr_do_stack_normalization;

extern char *vftr_start_date;
extern char *vftr_end_date;
extern bool in_vftr_finalize;

void vftr_initialize ();
void vftr_finalize () ;

void vftr_get_mpi_info (int *rank, int *size);

// test functions
int vftr_setup_test_1 (FILE *fp);
int vftr_setup_test_2 (FILE *fp);

#endif
