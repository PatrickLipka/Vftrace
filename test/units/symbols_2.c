#include "vftr_setup.h"
#include "vftr_symbols.h"
#ifdef _MPI
#include <mpi.h>
#endif

int main (int argc, char **argv) {

#if defined(_MPI)
  PMPI_Init(&argc, &argv);
  vftr_get_mpi_info (&vftr_mpirank, &vftr_mpisize);
#else 
  vftr_mpirank = 0;
  vftr_mpisize = 1;
#endif

  if (argc != 2) {
    printf ("Nr. of arguments: %d\n", argc);
    printf ("symbols_test_1: No input file\n", argc);
    return 1;
  }

  char *filename_in = argv[1];
  FILE *fp_in = fopen (filename_in, "r");
  if (fp_in == NULL) {
     printf ("File %s not found!\n", filename_in);
     return 1;
  }

  vftr_nsymbols = 0;
  vftr_get_library_symtab ("", fp_in, 0, 0);
  vftr_symtab = (symtab_t **) malloc (vftr_nsymbols * sizeof(symtab_t*));
  
  vftr_nsymbols = 0;
  rewind(fp_in);
  vftr_get_library_symtab("", fp_in, 0, 1);

  for (int i = 0; i < vftr_nsymbols; i++) {
    fprintf (stdout, "%s -> %s\n", vftr_symtab[i]->name, vftr_demangle_cpp(vftr_symtab[i]->name));
  }

  free (vftr_symtab);
  fclose(fp_in);
#ifdef _MPI
  PMPI_Finalize();
#endif
}
