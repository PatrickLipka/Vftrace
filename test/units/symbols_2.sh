#!/bin/bash
set -x
test_name=symbols_2
input_file=ref_input/symbols_2.x
output_file=symbols_2.out
ref_file=ref_output/symbols_2.out

if [ "x$HAS_MPI" == "xYES" ]; then
   ${MPI_EXEC} ${MPI_OPTS} ${NP} 1 ./${test_name} ${input_file} > ${output_file}
else
   ./${test_name} ${input_file} > ${output_file}
fi

if [ "$?" == "0" ]; then
  diff ${output_file} ${ref_file}
else
  exit 1
fi
