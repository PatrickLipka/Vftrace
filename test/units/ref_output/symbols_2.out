deregister_tm_clones -> deregister_tm_clones
register_tm_clones -> register_tm_clones
__do_global_dtors_aux -> __do_global_dtors_aux
frame_dummy -> frame_dummy
_Z41__static_initialization_and_destruction_0ii -> __static_initialization_and_destruction_0
_GLOBAL__sub_I__ZN10ANameSpace14InternalStruct17staticIntInternalE -> _GLOBAL__sub_I__ZN10ANameSpace14InternalStruct17staticIntInternalE
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EEC1ERKS6_ -> std::_Vector_base<..>::_Vector_base
_ZSt18uninitialized_copyIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPS5_ET0_T_SA_S9_ -> std::uninitialized_copy<..>
_ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EED1Ev -> std::vector<..>::~vector
_ZNSt16allocator_traitsISaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE37select_on_container_copy_constructionERKS6_ -> std::allocator_traits<..>::select_on_container_copy_construction
_ZSt7forwardIRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEOT_RNSt16remove_referenceIS8_E4typeE -> std::forward<..>
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EEC1EmRKS6_ -> std::_Vector_base<..>::_Vector_base
__libc_csu_fini -> __libc_csu_fini
_ZN10ANameSpace14InternalStruct6nested12nestedMethodESt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS8_EE -> ANameSpace::InternalStruct::nested::nestedMethod
_ZN9__gnu_cxx13new_allocatorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEC1ERKS7_ -> __gnu_cxx::new_allocator<..>::new_allocator
_ZSt8_DestroyIPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEvT_S7_ -> std::_Destroy<..>
_start -> _start
_ZN9__gnu_cxx13new_allocatorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE10deallocateEPS6_m -> __gnu_cxx::new_allocator<..>::deallocate
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE17_M_create_storageEm -> std::_Vector_base<..>::_M_create_storage
_ZN9__gnu_cxx13new_allocatorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEC1Ev -> __gnu_cxx::new_allocator<..>::new_allocator
_ZN9__gnu_cxxneIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIS6_SaIS6_EEEEbRKNS_17__normal_iteratorIT_T0_EESH_ -> __gnu_cxx::operator!=<..>
_ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EEC2ERKS7_ -> std::vector<..>::vector
_ZNK9__gnu_cxx17__normal_iteratorIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIS6_SaIS6_EEEdeEv -> __gnu_cxx::__normal_iterator<..>::operator*
_ZSt10_ConstructINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEJRKS5_EEvPT_DpOT0_ -> std::_Construct<..>
_ZNSaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEED2Ev -> std::allocator<..>::~allocator
_ZNSaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEC2ERKS5_ -> std::allocator<..>::allocator
_ZNSaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEC2Ev -> std::allocator<..>::allocator
_ZNSt20__uninitialized_copyILb0EE13__uninit_copyIN9__gnu_cxx17__normal_iteratorIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIS9_SaIS9_EEEEPS9_EET0_T_SI_SH_ -> std::__uninitialized_copy<..>::__uninit_copy<..>
_fini -> _fini
_ZN9__gnu_cxx13new_allocatorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEC2Ev -> __gnu_cxx::new_allocator<..>::new_allocator
_ZSt8_DestroyIPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES5_EvT_S7_RSaIT0_E -> std::_Destroy<..>
_ZN9__gnu_cxx13new_allocatorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEC2ERKS7_ -> __gnu_cxx::new_allocator<..>::new_allocator
_ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EED2Ev -> std::vector<..>::~vector
_ZNKSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE19_M_get_Tp_allocatorEv -> std::_Vector_base<..>::_M_get_Tp_allocator
_ZNKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE3endEv -> std::vector<..>::end
_ZnwmPv -> operator new
_ZN9__gnu_cxx13new_allocatorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEED2Ev -> __gnu_cxx::new_allocator<..>::~new_allocator
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE13_M_deallocateEPS5_m -> std::_Vector_base<..>::_M_deallocate
_ZNSt8ios_base4InitD1Ev@@GLIBCXX_3.4 -> std::ios_base::Init::~Init
_ZNKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE4sizeEv -> std::vector<..>::size
_ZNSt16allocator_traitsISaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE10deallocateERS6_PS5_m -> std::allocator_traits<..>::deallocate
_ZNKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE5beginEv -> std::vector<..>::begin
_ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EEC2ESt16initializer_listIS5_ERKS6_ -> std::vector<..>::vector
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE12_Vector_implD2Ev -> std::_Vector_base<..>::_Vector_impl::~_Vector_impl
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE19_M_get_Tp_allocatorEv -> std::_Vector_base<..>::_M_get_Tp_allocator
_ZN9__gnu_cxx13new_allocatorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE8allocateEmPKv -> __gnu_cxx::new_allocator<..>::allocate
_ZN10ANameSpace14InternalStruct6nested12nestedMethodEi -> ANameSpace::InternalStruct::nested::nestedMethod
_ZNK9__gnu_cxx17__normal_iteratorIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIS6_SaIS6_EEE4baseEv -> __gnu_cxx::__normal_iterator<..>::base
_ZNSaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEC1Ev -> std::allocator<..>::allocator
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EEC2EmRKS6_ -> std::_Vector_base<..>::_Vector_base
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EEC2ERKS6_ -> std::_Vector_base<..>::_Vector_base
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE12_Vector_implC2ERKS6_ -> std::_Vector_base<..>::_Vector_impl::_Vector_impl
_ZNKSt16initializer_listINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE5beginEv -> std::initializer_list<..>::begin
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE11_M_allocateEm -> std::_Vector_base<..>::_M_allocate
__libc_csu_init -> __libc_csu_init
_ZNKSt16initializer_listINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE4sizeEv -> std::initializer_list<..>::size
_ZN7OutsideC1Ei -> Outside::Outside
_ZSt10__distanceIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEENSt15iterator_traitsIT_E15difference_typeES9_S9_St26random_access_iterator_tag -> std::__distance<..>
_ZSt22__uninitialized_copy_aIN9__gnu_cxx17__normal_iteratorIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIS7_SaIS7_EEEEPS7_S7_ET0_T_SG_SF_RSaIT1_E -> std::__uninitialized_copy_a<..>
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE12_Vector_implD1Ev -> std::_Vector_base<..>::_Vector_impl::~_Vector_impl
_ZN9__gnu_cxx17__normal_iteratorIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIS6_SaIS6_EEEC2ERKS8_ -> __gnu_cxx::__normal_iterator<..>::__normal_iterator
_ZN9__gnu_cxx17__normal_iteratorIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIS6_SaIS6_EEEppEv -> __gnu_cxx::__normal_iterator<..>::operator++
_ZN7OutsideC2Ei -> Outside::Outside
_ZSt19__iterator_categoryIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEENSt15iterator_traitsIT_E17iterator_categoryERKS9_ -> std::__iterator_category<..>
_ZSt11__addressofINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEPT_RS6_ -> std::__addressof<..>
_ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EEC1ERKS7_ -> std::vector<..>::vector
_ZNSt16allocator_traitsISaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE8allocateERS6_m -> std::allocator_traits<..>::allocate
_ZN10ANameSpace14InternalStruct6nested12nestedMethodEv -> ANameSpace::InternalStruct::nested::nestedMethod
_ZNSt12_Destroy_auxILb0EE9__destroyIPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvT_S9_ -> std::_Destroy_aux<..>::__destroy<..>
_ZSt8distanceIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEENSt15iterator_traitsIT_E15difference_typeES9_S9_ -> std::distance<..>
_ZSt8_DestroyINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEvPT_ -> std::_Destroy<..>
_ZNSaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEED1Ev -> std::allocator<..>::~allocator
_ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EEC1ESt16initializer_listIS5_ERKS6_ -> std::vector<..>::vector
_ZNSt20__uninitialized_copyILb0EE13__uninit_copyIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPS7_EET0_T_SC_SB_ -> std::__uninitialized_copy<..>::__uninit_copy<..>
_ZSt18uninitialized_copyIN9__gnu_cxx17__normal_iteratorIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIS7_SaIS7_EEEEPS7_ET0_T_SG_SF_ -> std::uninitialized_copy<..>
_ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE19_M_range_initializeIPKS5_EEvT_SB_St20forward_iterator_tag -> std::vector<..>::_M_range_initialize<..>
_ZNK9__gnu_cxx13new_allocatorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE8max_sizeEv -> __gnu_cxx::new_allocator<..>::max_size
_ZN9__gnu_cxx13new_allocatorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEED1Ev -> __gnu_cxx::new_allocator<..>::~new_allocator
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE12_Vector_implC1ERKS6_ -> std::_Vector_base<..>::_Vector_impl::_Vector_impl
_ZNKSt16initializer_listINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE3endEv -> std::initializer_list<..>::end
_ZN10ANameSpace14InternalStruct6nested12nestedMethodENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE -> ANameSpace::InternalStruct::nested::nestedMethod
_ZdlPvS_ -> operator delete
__gxx_personality_v0@@CXXABI_1.3 -> __gxx_personality_v0@@CXXABI_1.3
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EED1Ev -> std::_Vector_base<..>::~_Vector_base
_ZN9__gnu_cxx17__normal_iteratorIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIS6_SaIS6_EEEC1ERKS8_ -> __gnu_cxx::__normal_iterator<..>::__normal_iterator
_ZNSaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEC1ERKS5_ -> std::allocator<..>::allocator
_ZNSt12_Vector_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EED2Ev -> std::_Vector_base<..>::~_Vector_base
_ZSt22__uninitialized_copy_aIPKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPS5_S5_ET0_T_SA_S9_RSaIT1_E -> std::__uninitialized_copy_a<..>
main -> main
_ZN9__gnu_cxx14__alloc_traitsISaINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEES6_E17_S_select_on_copyERKS7_ -> __gnu_cxx::__alloc_traits<..>::_S_select_on_copy
_init -> _init
